packages/loaders/src/zarr/pixel-source.ts

```diff
@@ -22,146 +23,203 @@ const DTYPE_LOOKUP = {
   i4: 'Int32'
 } as const;
 
-type ZarrIndexer<S extends string[]> = (
-  sel: { [K in S[number]]: number } | number[]
-) => number[];
-
-interface ZarrTileSelection {
-  x: number;
-  y: number;
-  selection: number[];
-  signal?: AbortSignal;
-}
-
-interface ZarrRasterSelection {
-  selection: number[];
-  signal?: AbortSignal;
-}
-
-interface Slice {
-  start: number;
-  stop: number;
-  step: number;
-  _slice: true;
+const X_AXIS_NAME = "x";
+const Y_AXIS_NAME = "y";
+const RGBA_CHANNEL_AXIS_NAME = "_c";
+
+type VivPixelData = {
+  data: zarr.TypedArray<Lowercase<SupportedDtype>>;
+  width: number;
+  height: number;
+};
+
+/**
+ * Error thrown when an assertion fails.
+ * Reference: https://github.com/hms-dbmi/vizarr/blob/862745c1c7c095748bbe97475da61807d5b49189/src/utils.ts#L399 
+ */
+export class AssertionError extends Error {
+  /** @param message The error message. */
+  constructor(message: string) {
+    super(message);
+    this.name = "AssertionError";
+  }
 }
 
-function slice(start: number, stop: number): Slice {
-  return { start, stop, step: 1, _slice: true };
+/**
+ * Make an assertion. An error is thrown if `expr` does not have truthy value.
+ * Reference: https://github.com/hms-dbmi/vizarr/blob/862745c1c7c095748bbe97475da61807d5b49189/src/utils.ts#L430
+ *
+ * @param expr The expression to test.
+ * @param msg The message to display if the assertion fails.
+ */
+export function assert(expr: unknown, msg = ""): asserts expr {
+  if (!expr) {
+    throw new AssertionError(msg);
+  }
 }
 
-type ZarrSource = Pick<ZarrArray, 'shape' | 'chunks' | 'dtype' | 'getRaw'>;
-
-class BoundsCheckError extends Error {}
-
-class ZarrPixelSource<S extends string[]> implements PixelSource<S> {
-  private _data: ZarrSource;
-  private _indexer: ZarrIndexer<S>;
+// Reference: https://github.com/hms-dbmi/vizarr/blob/main/src/ZarrPixelSource.ts 
```

Comment:
Is the plan to downstream this?

Reply:
Ideally, although the motivation behind this was more to get rid of the Zarr.js dependency in Vitessce.